scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -524,158 +525,174 @@
+                                 (->> matches
+                                      (map #(-> % :tasks vec))
+                                      flatten
+                                      (map #(-> % .getRequest :job))))
+        category->job-uuids (map-vals #(set (map :job/uuid %)) category->jobs)]
+    (log/debug "matched jobs:" (map-vals count category->job-uuids))
+    (when (not (empty? matches))
+      (let [matched-normal-jobs-resource-requirements (-> category->jobs :normal util/sum-resources-of-jobs)]
+        (meters/mark! matched-tasks-cpus (:cpus matched-normal-jobs-resource-requirements))
+        (meters/mark! matched-tasks-mem (:mem matched-normal-jobs-resource-requirements))))
+    category->job-uuids))
+
+(defn remove-matched-jobs-from-pending-jobs
+  "Removes matched jobs from category->pending-jobs."
+  [category->pending-jobs category->matched-job-uuids]
+  (let [remove-matched-jobs (fn remove-matched-jobs [category]
+                              (let [existing-jobs (category->pending-jobs category)
+                                    matched-job-uuids (category->matched-job-uuids category)]
+                                (remove #(contains? matched-job-uuids (:job/uuid %)) existing-jobs)))]
+    (pc/map-from-keys remove-matched-jobs (keys category->pending-jobs))))
+
+(defn- launch-matched-tasks!
+  "Updates the state of matched tasks in the database and then launches them."
+  [matches conn db driver fenzo fid]
+  (let [task-txns (for [{:keys [tasks leases]} matches
+                        :let [offers (mapv :offer leases)
+                              slave-id (-> offers first :slave-id :value)]
+                        ^TaskAssignmentResult task tasks
+                        :let [request (.getRequest task)
+                              task-id (:task-id request)
+                              job-id (get-in request [:job :db/id])]]
+                    [[:job/allowed-to-start? job-id]
+                     ;; NB we set any job with an instance in a non-terminal
+                     ;; state to running to prevent scheduling the same job
+                     ;; twice; see schema definition for state machine
+                     [:db/add job-id :job/state :job.state/running]
+                     {:db/id (d/tempid :db.part/user)
+                      :job/_instance job-id
+                      :instance/task-id task-id
+                      :instance/hostname (.getHostname task)
+                      :instance/start-time (now)
+                      ;; NB command executor uses the task-id
+                      ;; as the executor-id
+                      :instance/executor-id task-id
+                      :instance/slave-id slave-id
+                      :instance/ports (.getAssignedPorts task)
+                      :instance/progress 0
+                      :instance/status :instance.status/unknown
+                      :instance/preempted? false}])]
```

Comment:
Since this is a refactoring, it seems to me like this `for` is a good candidate for extracting into a function, perhaps:

```clj
(defn- task-txns [matches]
  (for ...))
```